### PR TITLE
Enable C++17

### PR DESCRIPTION
**Pull request by leoetlino**
C++17 allows for some nice cleanups and removes the need for standard library backports (optional/variant).

See discussion at https://github.com/dolphin-emu/dolphin/pull/6264#discussion_r158134178

### CODE REVIEW

File: Source/Core/Common/ChunkFile.h [hunk 1]
```diff
@@ -35,20 +35,8 @@
...
 // ewww
 
-#ifndef __has_feature
-#define __has_feature(x) (0)
-#endif
-
-#if (__has_feature(is_trivially_copyable) &&                                                       \
-     (defined(_LIBCPP_VERSION) || defined(__GLIBCXX__))) ||                                        \
-    (defined(__GNUC__) && __GNUC__ >= 5) || defined(_MSC_VER)
 #define IsTriviallyCopyable(T)                                                                     \
```
Comment (goldenercobra) [inline Source/Core/Common/ChunkFile.h]: In the spirit of cleanups, this should probably be a templated variable.

Comment (leoetlino) [reply Source/Core/Common/ChunkFile.h]: Done.

**Comment by lioncash** [comment]
Note that the MathUtil header `#include` directive can also likely be removed from source files that otherwise required it solely for MathUtil::Clamp in the past.

**Comment by Armada651** [comment]
Visual Studio also has an option for C++17 support, you should probably enable that one too: https://docs.microsoft.com/en-us/cpp/build/reference/std-specify-language-standard-version

**Comment by lioncash** [comment]
@Armada651 we already build with `/std:c++latest`, so the VS build will always be using the latest language standard.

File: Source/Core/Core/HW/VideoInterface.cpp [hunk 2]
```diff
@@ -327,7 +327,7 @@ void RegisterMMIO(MMIO::Mapping* mmio, u32 base)
             static_cast<u16>(1 +
                              m_HTiming0.HLW * (CoreTiming::GetTicks() - s_ticks_last_line_start) /
                                  (GetTicksPerHalfLine()));
-        return MathUtil::Clamp(value, static_cast<u16>(1), static_cast<u16>(m_HTiming0.HLW * 2));
+        return std::clamp(value, static_cast<u16>(1), static_cast<u16>(m_HTiming0.HLW * 2));
```
Comment (Warepire) [inline Source/Core/Core/HW/VideoInterface.cpp]: You could enforce this differently by using `std::clamp<u16>`

**Comment by Armada651** [comment]
@lioncash Right, I forgot to check the Common.props for it.

**Comment by leoetlino** [comment]
I think we'll need to wait for a while because old versions of Qt's `moc` (<5.9) chokes on C++17. And unfortunately, only Debian Sid and the next Ubuntu LTS (and rolling distros, of course) have 5.9.

Also, the Android, macOS, Debian and FreeBSD builders would need to be updated. It should be possible with Android and Debian, but I'm not sure about FreeBSD and macOS (delroth said even the latest SDK didn't have C++17 support?)

**Comment by lioncash** [comment]
> FreeBSD

Wouldn't endrift be able to update that whenever they have the time/want to update it?

> macOS 

macOS's version of clang in Xcode is based off clang 3.9 (based off the CMakeLists of Apple's version of clang on their open-source [page](https://opensource.apple.com/source/clang/clang-800.0.42.1/src/CMakeLists.txt.auto.html) (search for `LLVM_VERSION_MAJOR`), and this is only for 8.2 of the dev tools. It can be assumed that the latest version of Xcode (9.2), is some variant of clang 4, since it supports the -MJ and -Og, which were added in clang 4 ([changelog](http://releases.llvm.org/4.0.0/tools/clang/docs/ReleaseNotes.html#new-compiler-flags)). Their libc++ variant being used is also fairly up to date, as I've been able to use `<optional>` and `<string_view>` among other things with no issue on my MacBook in project that I've been working on.

For mostly (aside from one feature with partial support) full support of C++17, we would need to require clang 5, but for our purposes going that far wouldn't be necessary ([status page](https://clang.llvm.org/cxx_status.html))

> old versions of Qt's `moc` (<5.9) choke on C++17

Would you be able to elaborate a little more on this (or point me to a relevant place with more info)? As in, what way does it choke, just so I know a little more about the reason for it?

**Comment by leoetlino** [comment]
> Wouldn't endrift be able to update that whenever they have the time/want to update it?

I think so, turns out FreeBSD ports has clang 5.0.

For macOS, indeed, it's supposed to have clang 4.0, but last time delroth updated the guest (a week ago) he found out that `<optional>` wasn't present. Maybe it wasn't fully up-to-date?

>Would you be able to elaborate a little more on this (or point me to a relevant place with more info)? As in, what way does it choke, just so I know a little more about the reason for it?

moc is confused by the new namespace syntax ([QTBUG-59351](https://bugreports.qt.io/browse/QTBUG-59351)). That bug is fixed in 5.9.

**Comment by Warepire** [comment]
Side note: It's really easy to install Qt 5.9.x from PPAs on Ubuntu 16.04.

**Comment by orbea** [comment]
I would suggest sitting on this PR for a long while and let the C++17 support in compilers mature. Merging this now is just asking for compiler bugs...

For example see Mesen. https://github.com/SourMesen/Mesen/issues/305#issuecomment-355844660

**Comment by leoetlino** [comment]
Well, this is just about cleaning up the code by requiring C++17. Dolphin is already compiled in C++17 on Windows and on most Linux distributions (if the compiler supports it) since months and we haven't had any weird issue...

**Comment by sepalani** [comment]
Unless, we missed a strange UB like https://github.com/dolphin-emu/dolphin/pull/3143, I think it's fine.

**Comment by Tilka** [comment]
C++17 also improves template argument deduction:
```c++
std::mutex my_mutex;
std::unique_lock<std::mutex> my_lock(my_mutex);
```
can now be written as:
```c++
std::mutex my_mutex;
std::unique_lock my_lock(my_mutex);
```

**Comment by leoetlino** [comment]
Oh, indeed, I forgot about that one! However, because we can't merge this at the moment I think it would be wiser to make the changes later to avoid having tons of conflicts when this becomes mergeable.

**Comment by lioncash** [comment]
Updating this to notify that, in Apple's quest/race to the bottom when it comes to actual library support, Xcode 9.3 does not, in fact, move optional and other useful C++17 features out of the experimental namespace.

**Comment by leoetlino** [comment]
Ugh, that's annoying :/ with Ubuntu 18.04 releasing very soon, I was thinking April could be a good time to update the requirements, but that would force macOS users to install clang 5.0 with brew or something.

**Comment by leoetlino** [comment]
Is there any objection to merging this? It looks like all of our builders can now build this (even the macOS builder), and Ubuntu 18.04 ships with GCC 7.3.0. I would like to avoid having to rebase this more than 9 times :P

**Comment by jordan-woyak** [comment]
Do it! If there are actually any problems no one will fix them unless we force them to with this merge.

**Comment by BhaaLseN** [comment]
Go for it 💣💥

File: Source/CMakeLists.txt [hunk 3]
```diff
@@ -16,18 +16,13 @@ if(CMAKE_C_COMPILER_ID MATCHES "MSVC")
   check_and_add_flag(CPPLATEST /std:c++latest)
   check_and_add_flag(STANDARD_COMPLIANCE /permissive-)
 else()
-  # Enable C++17, but fall back to C++14 if it isn't available.
-  # CMAKE_CXX_STANDARD cannot be used here because we require C++14 or newer, not any standard.
+  # Enable C++17
+  # CMAKE_CXX_STANDARD cannot be used because it requires cmake 3.8+.
```
Comment (lioncash) [inline Source/CMakeLists.txt]: Overall looks good to me, but CMake's minimum should really be raised, in order to set the minimum through it, getting rid of the need to manually manage different compiler flags. Ubuntu 18.04 provides up to CMake 3.10 anyways.

Comment (leoetlino) [reply Source/CMakeLists.txt]: Done :)

**Comment by leoetlino** [comment]
With the latest commit, this doesn't build on FreeBSD anymore... The builder needs a CMake update as it currently uses CMake 3.8.

**Comment by endrift** [comment]
I've updated CMake to 3.14 on the FreeBSD builder.

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by leoetlino** [comment]
Thank you!

**Pull request merged**